src/vdc_collection.rs

```diff
@@ -1,197 +1,193 @@
     }
 
     /// Remove a credential from the store.
-    pub fn delete(&self, id: &str) -> Result<(), VdcCollectionError> {
-        match self.storage.remove(self.str_to_key(id)) {
-            Ok(_) => Ok(()),
-            Err(e) => Err(VdcCollectionError::DeleteFailed(e)),
-        }
+    pub fn delete(
+        &self,
+        key: Key,
+        storage: &Arc<dyn StorageManagerInterface>,
+    ) -> Result<(), VdcCollectionError> {
+        storage.remove(key).map_err(VdcCollectionError::from)
     }
 
     /// Get a list of all the credentials.
-    pub fn all_entries(&self) -> Result<Vec<String>, VdcCollectionError> {
-        let mut r = Vec::new();
-
-        match self.storage.list() {
-            Ok(list) => {
-                for key in list {
-                    let name = key.0;
-
-                    if name.starts_with(KEY_PREFIX) {
-                        if let Some(id) = name.strip_prefix(KEY_PREFIX) {
-                            r.push(id.to_string());
-                        }
-                    }
-                }
-            }
-            Err(e) => return Err(VdcCollectionError::LoadFailed(e)),
-        }
-
-        Ok(r)
+    pub fn all_entries(
+        &self,
+        storage: &Arc<dyn StorageManagerInterface>,
+    ) -> Result<Vec<Key>, VdcCollectionError> {
+        Ok(storage
+            .list()?
+            .iter()
+            .filter(|key| key.0.contains(KEY_PREFIX))
+            .map(ToOwned::to_owned)
+            .collect())
     }
 
     /// Get a list of all the credentials that match a specified type.
     pub fn all_entries_by_type(
         &self,
-        ctype: CredentialType,
-    ) -> Result<Vec<String>, VdcCollectionError> {
-        let mut r = Vec::new();
-
-        match self.all_entries() {
-            Ok(list) => {
-                for key in list {
-                    let cred = self.get(&key);
-
-                    if let Ok(Some(x)) = cred {
-                        if x.ctype == ctype {
-                            r.push(key);
-                        }
-                    }
-                }
-            }
-            Err(e) => return Err(e),
-        }
-
-        Ok(r)
+        ctype: &CredentialType,
+        storage: &Arc<dyn StorageManagerInterface>,
+    ) -> Result<Vec<Key>, VdcCollectionError> {
+        Ok(self
+            .all_entries(storage)?
+            .iter()
+            .filter(|key| key.0.contains(&Self::storage_prefix(ctype)))
+            .map(ToOwned::to_owned)
+            .collect())
     }
 
-    /// Convert a UUID to a storage key.
-    fn id_to_key(&self, id: Uuid) -> Key {
-        self.str_to_key(&id.to_string())
-    }
-
-    /// Convert a string ref to a storage key.
-    fn str_to_key(&self, id: &str) -> Key {
-        Key(format!("{}{}", KEY_PREFIX, id))
+    /// Return the storage prefix, given a credential type.
+    pub fn storage_prefix(ctype: &CredentialType) -> String {
+        format!("{KEY_PREFIX}{}.", ctype)
```

Comment:
I don't think we should add metadata about the credential to the key, it makes it more difficult to discover credentials, maintain and migrate at a later date.

Reply:
The reasoning here is to be able to ask the VdcCollection to list all credentials of a given type. If we store the type of the credential in the storage path, we can filter on that later, in `all_entries_by_type`.

Reply:
There's lots of things that we might need to filter on. If we need an effective way for lookup then we should maintain an external metadata map in my opinion.

Reply:
I mean, if I'm honest I think its unlikely that any individual device is going to store so many credentials that we can't just return an entire list and let the consumer deal with filtering it, but I'm not tied to this usage and I haven't had a reason to use it yet.

@Ryanmtate was this crucial for anything in particular that you want to speak up for it?